Augment `gen_server` timeout handling

Maria-12648430 opened the pull request:
While working on something entirely different, it ocurred to me that timeouts in `gen_server` have some drawbacks and pitfalls.

When the return of `gen_server:init` or a `handle_*` callback contains a timeout, the next iteration will wait for a message for that time, and if no message arrives, instantly call `handle_info` with the message `timeout` (ie, not _sending_ itself a `timeout` message).

For one, this convolution of `handle_info` for message and timeout handling makes it impossible to distinguish an actual timeout from the message `timeout` having been received. This means that you either have to make sure that, if you are using timeouts, such a message will not be sent to your `gen_server` implementation, or live with possible premature "fake" timeouts.

For another, it is very inconvenient if one wants to enricht the timeout with some supplementary data. This is only possible by putting it somewhere in the state, and must likely be removed from the state again if a message arrives before the timeout occurs.

Another option is to manually maintain timers, which is also inconvenient and requires quite some boilerplate code.

The changes in this PR allow `init` and `handle_*` callbacks to return the tuple `{timeout, Time, TimeoutMessage}`. In case a timeout occurs, `handle_info` will instantly be called with the message `{timeout, TimeoutMessage}`.

There are no tests or documentation yet, I first want to see if there is any interest in this feature at all.

RaimoNiskanen commented:
Hi!

You point out a bunch of design quirks/flaws with the `gen_server` time-out.  There are at least two more:
* Receiving a system message restarts the time-out, so it may become longer than requested.
* A time-out overrides the `hibernate_after` time-out, so the `gen_server` won't hibernate until after the time-out has elapsed.

But this PR only addresses one problem, namely attaching supplementary data to the time-out.

The PR's suggested form for the new time-out is `{timeout, Time, TimeoutMessage}` that looks exactly like `gen_statem`'s `{timeout, Time, EventContent}`, which makes it familiar.  That can be both good and bad...  Good because it is familiar, and bad because it still is subtly different from `gen_statem`'s time-out.

So that means that if this PR is accepted we have another flawed time-out in `gen_server`, and we cannot use that time-out form for a `gen_statem` time-out equivalent.  Not that we would want to (i think ever) do that.

I lean towards not augmenting the `gen_server` time-out.  It is what it is.  Augmenting it with `gen_statem` time-outs would not be a small rewrite.  If you want something better, use `gen_statem` or roll your own with `erlang` timers according to your current needs.

RaimoNiskanen commented on lib/stdlib/src/gen_server.erl:
```diff
@@ -2264,6 +2283,15 @@ loop(Parent, Name, State, CbCache, infinity, HibernateAfterTimeout, Debug) ->
             loop(Parent, Name, State, CbCache, hibernate, HibernateAfterTimeout, Debug)
     end;
 
+loop(Parent, Name, State, CbCache, {timeout, Time, TimeoutMsg} = Timeout, HibernateAfterTimeout, Debug) ->
```
Perhaps `Time =:= infinity` should be handled here instead of in `handle_msg/7`, `handle_common_reply/8` below, and two places above; `init_it/5` and `enter_loop/5` (missing).

Maria-12648430 replied:
TBH, I had it this way before, then thought having it in the `handle_msg` and friends functions would be better. But you're right, especially when combined with what @michalmuskala suggested. I'll change it back later.

RaimoNiskanen commented:
Then again, this particular enhancement doesn't change any of the good qualities about the `gen_server` time-out.  It is still simple and light-weight.

So the question is how much better would the `gen_server` time-out be with the addition of supplementary data...?  It might be worth it.

michalmuskala commented:
I mostly agree with what Raimo said, the only thing I would consider changing, if this is going forward would be to have the message delivered be just `Msg` for `{timeout, Time, Msg}` instead of `{timeout, Msg}`. If the user wants to wrap it, they can and it gives more flexibility if that's not necessary. It also gives a direct translation for the current option as `{timeout, Time, timeout}`

Maria-12648430 commented:
@RaimoNiskanen

> There are at least two more

I know, but I don't think that much can be done about those =( In any case, I wasn't trying to with this PR ;)

> Then again, this particular enhancement doesn't change any of the good qualities about the `gen_server` time-out. It is still simple and light-weight.
> 
> So the question is how much better would the `gen_server` time-out be with the addition of supplementary data...? It might be worth it.

To be sure, I took care not to change anything that would affect the current way how timeouts work in `gen_server`. Ie, anything that works now will keep working just the same with these changes applied. It also doesn't stand in the way of addressing the other 2+ quirks you mentioned, if we ever want to do that.

---

@michalmuskala 

> I mostly agree with what Raimo said, the only thing I would consider changing, if this is going forward would be to have the message delivered be just `Msg` for `{timeout, Time, Msg}` instead of `{timeout, Msg}`. If the user wants to wrap it, they can and it gives more flexibility if that's not necessary. It also gives a direct translation for the current option as `{timeout, Time, timeout}`

Excellent point, this is much better. Thanks for the suggestion :hugs:

---

I think I'll just let this discussion run for a while longer before making any changes to it, though :slightly_smiling_face:

RaimoNiskanen commented:
@Maria-12648430 wrote:
> It also doesn't stand in the way of addressing the other 2+ quirks you mentioned, if we ever want to do that.

But in a way it does.  If we would like to introduce the `gen_statem` time-outs in `gen_server`, that addresses most or all of the mentioned time-out issues, then this suggested format is one of the `gen_statem` formats.  So if we implement this PR, that `gen_statem` format has been occupied by a subtly different look-alike.

For instance; we could instead of this minimal suggestion, re-use that Timeout loop variable in `gen_server:loop` to be a (wrapped?) `reference()` to a running `erlang` timer.  I guess such an implementation could fix all `gen_server` time-out quirks except being possible to fake (but just like this suggestion decrease the impact of that quirk since the time-out message could be chosen arbitrarily).

Maria-12648430 commented:
@RaimoNiskanen ok, maybe I got you a bit wrong in the beginning.

So I understand that you are not against adding the _functionality_ this PR provides, but that you would like it to also address the other (at least two) quirks you mentioned, ie that adding only the ability to transmit timeout data falls short and may even get in the way of adressing the other issues later?

Tackling the other issues would mean changing `gen_server`s timeout handling to be more like how `gen_statem` handles event timeouts. Going this way is IMO not a bad thing, aligning them more closely, but also more effort. OTOH, looking at `gen_statem`, it is actually a pretty straightforward thing once you can see through all the layers of "`gen_statem` fancyness" around it. I think it can be done, and we would be willing to give it a try.

That is, unless you say that it's unlikely to be accepted, then of course we will save ourselves the trouble :stuck_out_tongue_winking_eye:

juhlig commented:
> we would be willing to give it a try

~What do you mean, "we"?~ Count me in 👍

RaimoNiskanen commented:
@Maria-12648430 Precisely.  I would like that when adding a time-out that looks like in `gen_statem`, it could be annoying if it doesn't behave like the same `gen_statem` time-out, which is the event time-out.

I would also like to not slow down `gen_server`, its engine keeps all its internal state in `gen_server:loop/7` variables, so if it turns out to be possible to implement an `erlang` timer based event time-out through re-use of the `Time` argument, then the engine should run as before.  Except that slightly more garbage would be produced if the callback module uses the new time-out.

I also would like that all quirks are addressed, not just the two I mentioned. But I guess it is impossible to completely fix fakeability without adding a new `gen_server` callback, so enabling a free choice of time-out message I'd say would be good enough.

So far I have not gotten any opinions from the rest of the team, I will have to get back to you about if it would be likely to be accepted...

Maria-12648430 commented:
> I also would like that all quirks are addressed, not just the two I mentioned.

Ok, well then... what _are_ the others? :sweat_smile:

Maria-12648430 commented:
> But I guess it is impossible to completely fix fakeability without adding a new `gen_server` callback, so enabling a free choice of time-out message I'd say would be good enough.

Indeed. In `gen_statem`, timeout events can be clearly distinguished from other events, external or internal. In `gen_server`, we would have to introduce a dedicated (albeit optional) callback like `handle_timeout/2`, which even with a fallback to `handle_info` would break any existing `gen_server` implementation which happens to have such a function (exported) already, which is not all that unlikely. The only way to mitigate this (that I can see) is to have another (optional) callback like what `callback_mode` does for `gen_statem`, which among other things could return how timeouts are to be handled (like, `{timeout, handle_info}` (current behavior, default) vs `{timeout, handle_timeout}`). Which AFAIK is a step on the road to over-engineering things, you are only a few steps away from just using `gen_statem` then (but on a side note, the "entirely different" thing I was working on as mentioned in the OP was the `supervisor` module, which is in turn a `gen_server`, which can't be easily exchanged for `gen_statem`).

> So far I have not gotten any opinions from the rest of the team, I will have to get back to you about if it would be likely to be accepted...

Ok, keep us in the loop :curly_loop: :slightly_smiling_face: :hugs:

RaimoNiskanen commented:
> > I also would like that all quirks are addressed, not just the two I mentioned.
>
> Ok, well then... what are the others? 😅

Hmmm, I had the feeling you mentioned a few, but it seems you mentioned supplementary data, fakeability, and that the `timeout` message is just passed in a call instead of sent.  My two points are just elaborations of your third.  So the known quirks should be:

* Fakeability (unfortunately it is not a word)
* No supplementary data
* Receiving a system message restarts the time-out
* A time-out overrides the hibernate_after time-out

So, as you point out, by enabling arbitrary time-out messages according to your suggestion, the two quirks I brought up would be the only remaining, and they are a consequence of the current time-outs not being delayed process messages.

RaimoNiskanen commented:
It is in general beneficial to reduce the number of loop arguments, because of the BEAM calling convention.  For example in `handle_msg/6` where `reply/2` is called (which calls `gen:reply/2` so it cannot be inlined), then 5..6 arguments have to be pushed onto the stack before the call and restored after since they are needed in the tail recursive call to `loop/7` that follows.  Reducing unnecessary register moves is desirable.

So keeping `Parent`, `Name`, `HibernateAfterTimeout` and probably `CbCache` in a record would probably be good, for performance and readability.  `CbCache` only changes after wakeup from hibernate so rebuilding the record then is probably not a problem.

`HibernateAfterTimeout` has to be looked up from the record almost every time the loop enters a `receive` statement.  A record lookup is slower than having it directly in argument, but that would probably not be possible to notice.

This is the kind of things I looked at when optimizing `gen_statem` to try to keep up with `gen_server`.  I haven't looked at optimizing `gen_server`, but there seem to be obvious things to do, like this.

For example I see that the argument order differs between `loop/7` and its helper functions such as `handle_common_reply/8,9` which should force a lot of register moves.   When the argument order is the same the compiler often figures out that it can just keep the argument registers where they are for the call, just having to move about a few.  But the code there introduces a new argument as the first instead of as the last.

Maybe helpers like `try_handle_continue/3` should be marked for inlining so calling them could avoid pushing all loop arguments.

One would have to look at the generated BEAM assembly to see these undesired register move clusters along the hot path that may be possible to avoid.  Finding a balance between keeping volatile state in argument registers versus avoiding lots of register shuffling is the game.

Maria-12648430 commented:
@RaimoNiskanen ok, then I guess I'll first make another PR to clean this up before seeing to the timeouts, it should make life a lot easier.

juhlig commented:
@RaimoNiskanen

@Maria-12648430 is busy this week, so I'm picking up the baton for a while.

We rebased this PR on current master (ie, with the changes from #9333) and reimplemented the timeout augmentation, trying to bring it closer to how `gen_statem` handles event timeouts.

The `init` and `handle_*` callbacks are now allowed to return as "actions":
* An integer `Time` ("plain" timeout), as before
* The atom `hibernate`, as before
* A tuple `{timeout, Time, TimeoutMsg}` (a plain timeout `Time` is the same as `{timeout, Time, timeout}`)
* A tuple `{hibernate, Boolean}` (the atom `hibernate` is the same as `{hibernate, true}`)
* A list consisting of the previous three (ie, _no_ plain timeouts), for example `[hibernate, {timeout, Time, TimeoutMsg}]`

The overall behavior is now as follows:
* A timeout of `infinity` is ignored.
* With a timeout of `0`, the `TimeoutMsg` is processed as the next event, ie without first receiving any other events from the message queue.
* All other timeout values start a timer, which sends the `TimeoutMsg` to itself. The timer ref is carried along, and if another message arrives before the timeout, it is cancelled and the message flushed from the message queue if it was already sent.
* By using a timer (vs a `receive` timeout), it is possible to send the `gen_server` into hibernation while it waits (with a combination of `hibernate` and `{timeout, ..., ...}`). Further, `system` messages do not cancel the timer.
* When the `gen_server` is instructed to hibernate, it will now first check if there is a message in the queue which would instantly wake it up again.

There are no new tests yet (but the current tests still pass) and the documentation has not been touched, either. We first want to hear what you think about it before progressing further :wink:

RaimoNiskanen commented:
The implementation looks complete and correct to me, you even handle zero time-out as anticipated.

However... I recognize that being able to combine `hibernate` with `{timeout, Time, Msg}` is a desirable feature.  But introducing a list of actions in the callback return timeout field might be to open a small can of worms.

In `gen_statem` an actions list was the only option since so many actions had to be possible to combine.

If we introduce a similar thing here; how similar should it be?  How dissimilar may it be?

In `gen_statem` the actions field action can be any single action, or a list of actions.  `Action` is equivalent to `[Action]`.  Here `Timeout` and `{continue, Term}` are not allowed in a list, `hibernate` is allowed as a single term or in the list, and `{timeout, Time, Msg}` and `{hibernate, Boolean}` are only allowed in a list.  This can of course be fixed, except for `{continue, Term}` that doesn't fit in with the others since it only makes sense alone.  I know `{continue, Term}` can be seen as not an action, but it occupies the same field as the actions, whereas in `gen_statem` it can only be `Action` or a list of `Action`s.

We also need to document how the actions are combined that is the last occurrence wins, since they set options for the `gen_server` loop.  The implementation seems to be consistent with `gen_statem` on this point, which I think it should be.  But it has to be documented nicely.

I don't know if it is a danger, or an opportunity, that we would get requests for more interesting actions, such as multiple replies or generic time-outs, if we introduce an actions list.  I lean towards that we should avoid adding such features to `gen_server` which maybe speaks against introducing an actions list.

Furthermore, having to create the `#actions{}` record in many places in the `gen_server` loop increases the amount of created garbage which may affect performance, also when not using the new `{timeout, Time, Msg}` feature.

A different possibility could be to have `{timeout, Time, Msg}` and `{timeout_hibernate, Time, Msg}`, or even `{hibernate, TimeoutTime, Msg}` as the only new timeout return field values.

A probably worse alternative would be `{timeout, Time, Msg, Options}` as that would conflict with the same action in `gen_statem` where `Options` is time-out options such as absolute time-out.  They may not collide, but having a hibernate time-out option in `gen_server` would make it missing in `gen_statem` (where it would interfere with the normal hibernate action).

So, should we have an actions list, or is there a better and slimmer alternative?

RaimoNiskanen commented:
Oops!  I noticed that this PR changes how the existing `gen_server` time-outs behave into them also being handled with a timer.

This is a subtle behaviour change that might bite us, and also a small performance regression...


I think the change to poll a message before going into hibernation is a good change.  The proposed code does the same after handling a system message if in hibernated state, which is something `gen_statem` doesn't do, but I think I will change that!

juhlig commented:
@RaimoNiskanen it's up to you and your colleagues to decide, we can change it any way you want, give or take :)

RaimoNiskanen commented:
@juhlig: Do you have any thoughts about actions list vs. `{timeout, T, M}` and `{hibernate, T, M}`?

juhlig commented:
@RaimoNiskanen having only single actions is ok I think. The initial proposal @Maria-12648430 made was only about extending the plain timeout with a message, so I'm sure she won't mind, either. And at the end of the day, it doesn't block the way to having action lists later if we find that we need them after all.

That said, the particular form `{hibernate, T, M}` looks a little weird to me. It may be misleading, in the way that users may think of it like "hibernate after `T` "(framed like that, the `M` makes no sense, but still).

> I noticed that this PR changes how the existing `gen_server` time-outs behave into them also being handled with a timer.
> 
> This is a subtle behaviour change that might bite us, and also a small performance regression...

~Whatever could go wrong?~ Well, I'm not sure if this could be a problem. All that changes in general behavior is that timeouts now "survive" system messages and hibernation. But then again, of course we don't know everything that exists out there and what subtle changes may eventually break :man_shrugging:

Regarding the performance regression, I would argue that if a user provided a timeout, it means that he is willing to wait for a while. Also, the timeout values `infinity` and `0` (see below) are treated specially and don't start a timer.

While we are at subtle changes, there is another: a zero timeout will now process the timeout message instantly, as if it was the next event, without receiving any other messages from the queue. This is pretty much what `gen_statem` does, but different from what `gen_server` did in the old implementation. OTOH, this remedies an old issue ("do not rely on zero timeouts for async continuation: someone _could_ have sent a message") which was ultimately worked anound with `handle_continue`. OTOH, it _is_ a behavior change.

Now, it _is_ possible to circumvent the above behavior changes by treating "plain" timeouts (ie, given as just an integer or `infinity` vs the `{timeout, T, M}` feature) specially, "the old way". This distinction may be confusing to users, though.

michalmuskala commented:
> While we are at subtle changes, there is another: a zero timeout will now process the timeout message instantly, as if it was the next event, without receiving any other messages from the queue. This is pretty much what `gen_statem` does, but different from what `gen_server` did in the old implementation. OTOH, this remedies an old issue ("do not rely on zero timeouts for async continuation: someone _could_ have sent a message") which was ultimately worked anound with `handle_continue`. OTOH, it _is_ a behavior change.

I've seen timeouts used to implement some sort of "out of band" work solution. If there's no actual jobs waiting for processing, the server does non-urgent maintenance work - this being implemented with 0 timeout. With the change as you describe, this way of using timeouts would break.

RaimoNiskanen commented:
@juhlig Wrote:
> Now, it is possible to circumvent the above behavior changes by treating "plain" timeouts (ie, given as just an integer or infinity vs the {timeout, T, M} feature) specially, "the old way". This distinction may be confusing to users, though.

Well spotted!  And @michalmuskala gives another reason I couldn't think of.  We need to be aggressively backwards compatible with something as central as `gen_server`, so the old time-outs should work exactly as quirky as before.  And the new zero time-out should work like in `gen_statem`.

I agree on your woes about `{hibernate, T, M}`, but it is short and no introduces no new tag to remember.  How about my other suggestion `{hibernate_timeout, T, M}`, or `{timeout_hibernate, T, M}`? (`timeout_and_hibernate`, ...)

juhlig commented:
> We need to be aggressively backwards compatible with something as central as `gen_server`, so the old time-outs should work exactly as quirky as before. And the new zero time-out should work like in `gen_statem`.

Ok, can do :+1: 

> I agree on your woes about `{hibernate, T, M}`, but it is short and introduces no new tag to remember. How about my other suggestion `{hibernate_timeout, T, M}`, or `{timeout_hibernate, T, M}`? (`timeout_and_hibernate`, ...)

Of the suggested options, personally I would pick `timeout_and_hibernate`, since this wording sort of "separates" `timeout` and `hibernate`.

One other thought that I forgot to mention: `gen_statem` has the option to use timeouts with options (`{timeout, T, M, Opts}`), where `Opts` can be `{abs, ...}`. Should we have that in `gen_server`, too?

RaimoNiskanen commented:
Great!  Let's go with `timeout_and_hibernate` for now...

Timeout options would be useful (absolute time-outs), I don't see why we shouldn't add them, making this time-out equivalent to `gen_statem`'s event time-out.  The options list can of course be added later, but why not right from the start.

RaimoNiskanen commented:
I am having second thoughts about not going into hibernation when a message poll gets a message...  That change is probably not as innocent as it might seem.

A busy server that decides to hibernate for some reason, might depend on hibernation to clean up memory usage and minimize the heap.  A similar case was discussed during the design of `gen_statem`, that had to handle the case when a callback module requests hibernation, but there are postponed events to process.  We decided to do a garbage collection since that might be an essential side effect.

So, in this case, if we should do such an optimization, we should recognize when it hits, and do a garbage collection.  Or we should not do the optimization.

juhlig commented:
@RaimoNiskanen WDYT of #9406? Should we use it here?

juhlig commented:
@RaimoNiskanen last commit removes action lists for single actions, restores the original behavior in case of a plain timeout, and also removes the optimization of skipping hibernation.

juhlig commented:
> So, in this case, if we should do such an optimization, we should recognize when it hits, and do a garbage collection. Or we should not do the optimization.

I checked [what `gen_statem` does in such a case](https://www.erlang.org/doc/apps/stdlib/gen_statem.html#t:hibernate/0), and turns out it does not go into hibernation but simulates it by calling `erlang:garbage_collect/0` instead. So I think we should do the same in `gen_server`.

juhlig commented:
Not sure what's wrong with the failing check, it passes locally :man_shrugging:

Maria-12648430 commented:
@RaimoNiskanen WDYT of @juhlig's implementation as it stands now? If there are no more objections regarding the functionaliyt, I would start writing tests and documentation.

IngelaAndin commented:
To me hibernate_with_timeout conveys the intent better.

Maria-12648430 commented:
@IngelaAndin 

> To me hibernate_with_timeout conveys the intent better.

Hm... I think I disagree here :sweat_smile: To me, the term "with" conveys the notion that the two things, namely "hibernate" and "timeout", are interdependent or connected (a possible interpretation that comes to mind is "hibernate after timeout"). But in fact, they are two distinct things :woman_shrugging:

IngelaAndin commented:
I misunderstood the timeout_hibernate to be a hibernare_after also when saying timeout_and_hibernate it feels unclear that hibernate is something that happens right away and the timeout will wake up the process unless some other message comes first in which case the timer will be canceled.

IngelaAndin commented:
Maybe {wakeup_hibernation,  Time} ?

RaimoNiskanen commented:
@juhlig: I think PR #9406 could be used here to simplify code, but we would have to wait for it to be merged to master and then rebase.

@Maria-12648430: I found a few flaws in @juhlig:s current proposal, and then got a bit carried away... :-/ 

So I created PR #9466 that adds commits on top of this PR that may be picked or used for inspiration.

#### First commit
* When jumping from `loop/5` to `decode_msg/6` after emulated hibernation the `Hib` argument should be `true`.  This is because the message that prevented hibernation could be a system message and then we should pretend it didn't happen and hibernate again.  We emulate that system messages "doesn't" break hibernation.
* `extract_action/1` should handle multiple options (`{abs,Bool}` options) correctly as in the last wins.
* Rewrote `extract_action/1` to produce less garbage.
* Refactored `decode_msg/6` to have an exit function `decode_msg/4` common to two clauses which made the code shorter and clearer, I think.
* Changed `maybe` constructs that had only one `?=` operator since such are just a `case` in disguise.  A style question where I think `case` conveys the intention better.
* Wrote more explicit error handling for `extract_action/1` in `enter_loop/5`.
* Added missing handling of `{timeout_and_hibernate, 0, ...}` which is a stupid combination but it should work.

#### Second commit
* Divided `loop/6` into `loop/5` without the not needed `TRef` argument and the tail of `loop/6`.

#### Third commit
* Refactored `extract_action/1` to `handle_action/1` that mostly does the action and returns a more directly usable value ("`LoopAction`") which simplifies `loop/5`.  This produces more garbage, but just a little.
* Removed guard matchings of functions that return a constant atom, which is according to Dialyzer:s expectations.
* Removed a redundant clause for `{continue, _}` in `handle_common_reply/5`.
* Moved `handle_action/1` below `handle_common_reply/*` according to the module's general convention that called functions are after the call.

RaimoNiskanen commented:
Regarding the name for `hibernate` combined with `{timeout, T, M}`.

I agree with @IngelaAndin that `timeout_and_hibernate` doesn't indicate well that hibernation happens immediately.

My first proposal was simply `{hibernate, T, M}`, but @juhlig argued that it could mean hibernate after the time `T`.

I think that @IngelaAndin's proposal `hibernate_with_timeout` indicates that we will hibernate (immediately), but there is a time-out.

Wording is hard; I cannot come up with a clearer proposal right now.

`hibernate_until`?

juhlig commented:
> If nothing better is proposed I think we should go with `{hibernate, T, M}` that I think indicates hibernation, but with a time `T`, and if it isn't obvious what `T` means the documentation will tell.

There seems to be no short term that conveys it all, so yeah, let's go with `{hibernate, ...}` and document it well. We will have to document the subtly different behaviors of plain vs "new" timeouts, anyway.

RaimoNiskanen commented:
@Maria-12648430: So, what do you say about my suggested fixes in PR #9466?  Will you pick the parts you like, or should I push them all onto this PR?

Maria-12648430 commented:
@RaimoNiskanen will take a closer look on monday 🙃

Maria-12648430 commented:
@RaimoNiskanen looks like I'll be busy next week too, so I took that closer look at your suggestions today :wink: Aside from what I think is an artifact carried over from an earlier commit, I think they are good and sound :+1: I have nothing to specially pick, or rather to leave out, so I think it would be best to just push them on this one if that is ok for you.

RaimoNiskanen commented:
I also pushed the rename to `{hibernate, T, M [,Opts]}`.

I think we have agreement on the functionality.  Test cases, type fixes and documentation are missing.  In particular; should `{continue,_}` be a subtype of `action()`..., should we export `action()`...?

michalmuskala commented on lib/stdlib/src/gen_server.erl:
```diff
@@ -2243,33 +2243,62 @@ init_it(Mod, Args) ->
+            handle_common_reply(ServerData, State, Msg, From, Reply, Debug1)
     end;
+%%
+loop(ServerData, State, LoopAction, Debug) ->
+    case LoopAction of
+        {timeout_zero, TimeoutMsg} ->
+            decode_msg(ServerData, State, TimeoutMsg, undefined, Debug, false);
+        {timeout, TRef, HibInf} ->
+            loop(ServerData, State, HibInf, TRef, Debug);
+        HibT ->
+            loop(ServerData, State, HibT, undefined, Debug)
+    end.
```
This will do some argument shuffling since `Debug` goes from 4th to 5th argument. If the `LoopAction` and `Debug` order was flipped, this wouldn't be necessary.

RaimoNiskanen replied:
That could be worth changing.  It also hits `loop/5`, but I don't find any call that goes against that argument order.

michalmuskala commented:
As a general point - is this change big enough to warrant an EEP?

RaimoNiskanen commented on lib/stdlib/src/gen_server.erl:
```diff
@@ -2292,27 +2321,33 @@ update_callback_cache(#server_data{module = Mod} = ServerData) ->
-            Debug1 = sys:handle_debug(Debug, fun print_event/3,
-                                      ServerData#server_data.name, {in, Msg}),
-            handle_msg(ServerData, State, Msg, Debug1)
+	{timeout, TRef, Msg1} when TRef =/= undefined ->
```
The tagging is "risk free" because `TRef` ensures that it is the timeout message we are expecting.  The `timeout` tag is only valuable for debugging, and might  optimize pattern matching since most incoming messages will be found to not match by atom comparison with `timeout`.

The receive is not eligible for the reference receive optimization because it is a wildcard receive.  All messages are received without pattern matching and then processed.  So the mailbox is not scanned - the first message is picked.

Maria-12648430 replied:
> Also - is this receive optimised? We now have a pattern with a pre-bound variable `TRef` - we don't want to scan the entire mailbox in gen_servers

This is a `case`, not a `receive`.

michalmuskala replied:
You are completely right. I somehow missed both of those details, and importantly the fact that this is a case 🤦‍♂️. I'm sorry for the noise

Maria-12648430 replied:
> In `gen_server` it is trickier since we want to not increase the size of server state data (too much), but I am thinking about using a server specific reference to tag the messages.

That, and a counter that gets incremented whenever a timeout timer gets started and again when it gets canceled or the message is received. We send that counter value in the timer message and also carry it along in a loop variable. Any such server-ref tagged timeout message carrying a counter value lower (or, even more generally, _any other_) than what is carried in the loop when it is received can be ignored as stale. So if a timer can't be canceled because it has already "fired", there is no need to flush the message out with a selective receive. WDYT @RaimoNiskanen?

Maria-12648430 replied:
I remember @juhlig once told me there is a limit beyond which handling integers gets slower, 2^61 I think it was. I don't know if that is still the case, but even if: given that timers have millisecond resolution, it would take ~73 million years until the counter would exceed this value with continuous 1ms timeouts. Zero relative timeouts are treated specially and bypass timers. The only uncertainty are absolute timeouts which are set to a time in the past.

juhlig replied:
Hm, what do you need that counter for? The timer ref is already carried around and unique. We only need an additional way to identify timeout messages as originating from the gen_server, which a server-specific ref would be for. This could be carried around in the `#server_data`, it never changes for the lifetime of the process.

RaimoNiskanen replied:
@juhlig: You have read my mind, to the letter.  @Maria-12648430 : I think that would be sufficient.

The time-out message can be `{timeout, TRef, ServerRef}`, what we store in the current time-out ref variable can be `[TRef|Msg]`, and in the `#server_data` add a field for `ServerRef`.

`gen_statem` stores the timer message in its server state.  To store the timer message in the timer has pros and cons.  On the plus side is that the message (that may be large) eludes garbage collections, which gets the negative side that the content has to be copied there and back.  Nevertheless there may be some value in both `gen_server` and `gen_statem` using the same storage strategy.

juhlig replied:
@RaimoNiskanen heh 😅

I already pushed a suggestion before I read your comment. Please take a look.

A nice side effect is that now we can use async when cancelling the timer and so don't have to wait.

Maria-12648430 replied:
@juhlig @RaimoNiskanen right, what was I thinking :sweat_smile:

juhlig replied:
@RaimoNiskanen ah, reading it a second time, I realize that you had a somewhat different way in mind. Makes sense I think, I'll adapt it later today.

juhlig replied:
Done, WDYT?

RaimoNiskanen replied:
Looks great!

I made a small suggestion that maybe is a clarification...

RaimoNiskanen commented:
@michalmuskala: I think this addition is small enough to not warrant an EEP, but I will not take a fight for that.

Maria-12648430 commented:
> I also pushed the rename to `{hibernate, T, M [,Opts]}`.

:+1:

> I think we have agreement on the functionality. Test cases, type fixes and documentation are missing.

I'll work on that soon, but as I'm currently busy and @juhlig is out sick, it may take a little :woman_shrugging:

> In particular; should `{continue,_}` be a subtype of `action()`...,

By intuition: yes.

> should we export `action()`...?

`gen_statem` exports `action`, so I suggest we do the same here.

Maria-12648430 commented:
@RaimoNiskanen I put `{continue, ...}` under `actions()` and made some first tests in the last commits.

RaimoNiskanen commented:
@Maria-12648430: And you rebased to a later 'master', one that has got `hibernate/0`...  It is tempting to use it

Maria-12648430 commented:
> @Maria-12648430: And you rebased to a later 'master', one that has got `hibernate/0`... It is tempting to use it

Oh, all riiight 😍

Maria-12648430 commented:
@RaimoNiskanen #9509 --> this breaks the tests. It could be worked around, but I consider it a bug and will wait for a fix.

RaimoNiskanen commented on lib/stdlib/src/gen_server.erl:
```diff
@@ -2292,27 +2305,39 @@ update_callback_cache(#server_data{module = Mod} = ServerData) ->
+	{timeout, TRef, Tag} when TRef =:= hd(Timer) ->
+            decode_msg(ServerData, State, tl(Timer), Debug);
+	{timeout, _Stale, Tag} ->
+	    LoopAction = case Hib of
+			     true -> hibernate;
+			     false -> infinity
+			 end,
+	    loop(ServerData, State, LoopAction, Debug, Timer);
```
Maybe make this a new function `system_continue(ServerData, State, Hib, Debug, Timer)`  split out from `system_continue/3` to show that this goes the same way as when handling (ignoring) a system message.  Or it can be a function `loop_continue(ServerData, State, Hib, Debug, Timer)` that is also called from `system_continue/3`.

Maria-12648430 replied:
@RaimoNiskanen done, I went for `loop_continue` :)

RaimoNiskanen commented on lib/stdlib/src/gen_server.erl:
```diff
@@ -2417,132 +2442,183 @@ try_terminate(#server_data{module = Mod}, State, Reason) ->
-system_continue(Parent, Debug, [#server_data{parent=Parent} = ServerData, State, Time]) ->
-    loop(update_callback_cache(ServerData), State, Time, Debug).
+system_continue(Parent, Debug, [#server_data{parent=Parent} = ServerData, State, Timer, Hib]) ->
+    Action = case Hib of
+		 true -> hibernate;
+		 false -> infinity
+	     end,
+    loop(update_callback_cache(ServerData), State, Action, Debug, Timer).
```
In connection to my suggestion for `decode_msg/6` above:

``` erlang
    system_continue(update_callback_cache(ServerData), State, Hib, Debug, Timer).

system_continue(ServerData, State, Hib, Debug, Timer) ->
    Action = case Hib of
￼		 true -> hibernate;
￼		 false -> infinity
￼	     end,
￼    loop(ServerData, State, Action, Debug, Timer).
```

Maria-12648430 commented:
> @RaimoNiskanen #9509 --> this breaks the tests. It could be worked around, but I consider it a bug and will wait for a fix.

Ok, @jhogberg says this is not a bug, and so I will go with the workaround solution for testing.

RaimoNiskanen commented on lib/stdlib/src/gen_server.erl:
```diff
@@ -2243,33 +2232,73 @@ init_it(Mod, Args) ->
+
+loop_hibernate(ServerData, State, Debug, Timer) ->
+    erlang:hibernate(),
+    loop_wakeup(ServerData, State, Debug, Timer).
+
+loop_wakeup(ServerData, State, Debug, Timer) ->
```
I don't see any point in `loop_wakeup/4` vs. inlining it in `loop_hibernate/4`.

RaimoNiskanen replied:
I like the new `hibernate/0` and how the code using it becomes simpler - it is a more convenient concept!

Maria-12648430 replied:
For the `gen_server` code by itself, yes (and I had it this way at first). But it may hide a bug: in the tests, checking if a process is in hibernation is done by comparing the result of `erlang:process_info(Pid, current_function)` against `{gen_server, loop_hibernate, 4}`. This result is returned both when the process _is_ in hibernation (ie, in `erlang:hibernate/0`), but _also_ if it is _not_ hibernating any more but instead waiting in the following `receive` (which would be a bug).

I agree that this should not happen, in that if a process gets "past" the `erlang:hibernate/0`, it _must_ be because there is a message. Not sure we should cover this here though, because if that ever happens, the bug would not be in `gen_server`.

RaimoNiskanen replied:
Right, I see!  That might prove a useful debugging property one day.

RaimoNiskanen commented:
@Maria-12648430 and @juhlig: What do you think about changing the non-existing timer from `undefined` to `[]`?  To me it _maybe_ feels better to have he same basic type.  One could envision a future type checker / linter to think it is bad style to do `hd(undefined)` although in a guard.

Just a thought...

Maria-12648430 commented:
> @Maria-12648430 and @juhlig: What do you think about changing the non-existing timer from `undefined` to `[]`? To me it _maybe_ feels better to have he same basic type. One could envision a future type checker / linter to think it is bad style to do `hd(undefined)` although in a guard.
> 
> Just a thought...

@RaimoNiskanen Sure, see last commit :wink:

Maria-12648430 commented:
@RaimoNiskanen I added some more tests. Actions returned from `init`, `handle_call`, `handle_cast` and `handle_info` as well as passed to `enter_loop` should now be covered.

Maria-12648430 commented:
@RaimoNiskanen I added some more tests to also cover timeout options. As those tests become more numerous, the time they take gets noticeable...

I also noticed that we only provide for timeout options, that is, a list thereof (like `[{abs, ...}]`). However, `gen_statem` also allows for single options (like `{abs, Abs}`). I added that here, too.

RaimoNiskanen commented on lib/stdlib/src/gen_server.erl:
```diff
@@ -2568,10 +2580,13 @@ handle_timeout(ServerData, T, M, HibInf) ->
     end.
 
-handle_timeout(ServerData, T, M, HibInf, [], Abs) when ?is_timeout(Abs, T) ->
+handle_timeout(ServerData, T, M, HibInf, {abs, Abs}, _Abs) when is_boolean(Abs),
+								?is_timeout(Abs, T) ->
```
This opens up for an improper list with `{abs, Abs}` in the tail.

To not accept improper lists I tend to use this pattern:
``` erlang
handle_action(ServerData, Action) ->
    :
        {timeout, T, M, Opts} ->
            handle_timeout(ServerData, T, M, infinity, listify(Opts), false);
    :
    end.

-compile({inline, [listify/1]}).
listify(Item) when is_list(Item) ->  Item;
listify(Item)                    -> [Item].
```

RaimoNiskanen commented:
@Maria-12648430 I added single options to `gen_statem` because that all started with the callback return value in `gen_server` that has a single field `Timeout` and it was desired that the same style would work from a `gen_statem` callback.  After that all other places in `gen_statem` where it is possible to have either a single option or a list, allows it.

Within `gen_server` there is no such precedence.  But to make the time-out field as similar as possible as in `gen_statem`, I agree that we should allow single options here too, as you suggest...

Maria-12648430 commented:
@RaimoNiskanen ok, added prevention of improper options lists as suggested :+1:

juhlig commented:
@RaimoNiskanen I was just getting back to this, and noticed something. If I'm not mistaken, in the old version, a `system` message would _restart_ a plain timeout. In the new version, I think a `system` message would _cancel_ a plain timeout?

RaimoNiskanen commented:
I hope this fixes that issue:
https://github.com/RaimoNiskanen/otp/tree/gen_server_timeout

juhlig commented:
> I hope this fixes that issue: https://github.com/RaimoNiskanen/otp/tree/gen_server_timeout

@RaimoNiskanen Proofreading and a quick test says it does :+1:

RaimoNiskanen commented:
Great to hear!

When you and @Maria-12648430 continue with what's missing (was it documentation and I don't know if the type specs were completed), can you start with cherry-picking this bugfix?

juhlig commented:
> When you and @Maria-12648430 continue with what's missing (was it documentation and I don't know if the type specs were completed), can you start with cherry-picking this bugfix?

@RaimoNiskanen I did the fix now in #9615 (basically, what you provided plus a test). Maria is currently n/a, and I'm not sure how long documentation and spec refining will really take for me.

RaimoNiskanen commented:
Excellent!  Thank you for writing a test.

This fix should be in 28.0-rc3 that have code freeze the first week in April.  I will add PR #9615 to our daily tests for that.

Doc freeze for 28.0 is the first days of May.  Do you think she will have time to work on the documentation and specs, or would it be better if I get on with it?

juhlig commented:
> Doc freeze for 28.0 is the first days of May. Do you think she will have time to work on the documentation and specs, or would it be better if I get on with it?

I really can't say about her, it's complicated let's say 😕 That's why I have been working on it as stand-in, even though I'm notoriously bad at explaining stuff.

Anyway, it should be doable I think, made some headway at least. If I don't get it done by next week, say, I'll let you know ok?

RaimoNiskanen commented:
I have written the documentation, I think, but it is probably sloppy since it went rater fast.  I will let it brew overnight, and push tomorrow.

Suddenly, I realized that the new time-outs also should handle `Time = infinity`, as `gen_statem` does, for completeness.  I have a fix, but so far no test case.  Push tomorrow...
